File: client_app/lib/models/paginated_response.dart
class PaginatedResponse<T> {
  final bool success;
  final String message;
  final List<T> data;
  final int pageNumber;
  final int pageSize;
  final int totalElements;
  final int totalPages;
  final bool first;
  final bool last;
  final String? timestamp;

  PaginatedResponse({
    required this.success,
    required this.message,
    required this.data,
    required this.pageNumber,
    required this.pageSize,
    required this.totalElements,
    required this.totalPages,
    required this.first,
    required this.last,
    this.timestamp,
  });

  factory PaginatedResponse.fromJson(
    Map<String, dynamic> json,
    T Function(Map<String, dynamic>) fromItem,
  ) {
    final rawData = json['data'] as List<dynamic>? ?? [];
    return PaginatedResponse(
      success: json['success'] ?? false,
      message: json['message'] ?? '',
      data: rawData.map((e) => fromItem(e as Map<String, dynamic>)).toList(),
      pageNumber: json['pageNumber'] ?? 0,
      pageSize: json['pageSize'] ?? 10,
      totalElements: (json['totalElements'] ?? 0) is int
          ? json['totalElements']
          : (json['totalElements'] as num).toInt(),
      totalPages: json['totalPages'] ?? 0,
      first: json['first'] ?? true,
      last: json['last'] ?? true,
      timestamp: json['timestamp'],
    );
  }
}

File: src/main/java/com/example/employeeapi/dto/PaginatedResponse.java
package com.example.employeeapi.dto;

import com.fasterxml.jackson.annotation.JsonInclude;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.time.LocalDateTime;
import java.util.List;

@Data
@NoArgsConstructor
@AllArgsConstructor
@JsonInclude(JsonInclude.Include.NON_NULL)
public class PaginatedResponse<T> {

    private boolean success;
    private String message;
    private List<T> data;
    private int pageNumber;
    private int pageSize;
    private long totalElements;
    private int totalPages;
    private boolean first;
    private boolean last;
    private LocalDateTime timestamp = LocalDateTime.now();

    public PaginatedResponse(boolean success, String message, List<T> data,
                             int pageNumber, int pageSize, long totalElements,
                             int totalPages, boolean first, boolean last) {
        this.success = success;
        this.message = message;
        this.data = data;
        this.pageNumber = pageNumber;
        this.pageSize = pageSize;
        this.totalElements = totalElements;
        this.totalPages = totalPages;
        this.first = first;
        this.last = last;
        this.timestamp = LocalDateTime.now();
    }
}
